### PR TITLE
DEPS: drop Python 3.7 and NumPy 1.17 from tests (#41678)

### LINKED ISSUE #41678: Drop support for Python 3.7
from https://numpy.org/neps/nep-0029-deprecation_policy.html

> On Dec 26, 2021 drop support for Python 3.7 (initially released on Jun 27, 2018)

The next release after 1.3 can be scheduled for 6 months, and we can drop 3.7 support in the next release.

### PR DESCRIPTION
**Pull request by fangchenli**
closes #41678

Most of the new minimum support versions were released between Oct 14, 2019, Python 3.8's initial release date, and June/July 2020.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @fangchenli for getting this started. To close the issue, are you planning to update the release notes in this PR too?

**Comment by fangchenli** [comment]
> Thanks @fangchenli for getting this started. To close the issue, are you planning to update the release notes in this PR too?

I'll update the doc after at last.

File: ci/deps/actions-38-minimum_versions.yaml
Comment (fangchenli) [inline ci/deps/actions-38-minimum_versions.yaml:21]: Conda cannot resolve `1.18.3`. so use `1.18.5` instead.

**Comment by MarcoGorelli** [comment]
could update this to `--py38-plus`

https://github.com/pandas-dev/pandas/blob/0b68d87a4438a13f14a2ed5af2e432df02eb0b2c/.pre-commit-config.yaml#L59-L63

**Comment by simonjayhawkins** [comment]
> could update this to `--py38-plus`

I think maybe code changes should be a follow-on. This PR should probably just be the ci setup and docs.

**Comment by fangchenli** [comment]
There are 17 failures in Python 3.9 Windows build relate to `pyarrow`.

Most of them are:
`pyarrow.lib.ArrowNotImplementedError: Support for codec 'snappy' not built`

**Comment by simonjayhawkins** [comment]
> update the release notes in this PR too?

I should have said install.rst too.

I think it would be easier for reviewers to see the significant changes, but I also appreciate that while getting things working is an additional burden, so add when ready.

File: pandas/tests/computation/test_eval.py
Comment (simonjayhawkins) [inline pandas/tests/computation/test_eval.py:1286]: could maybe leave these changes for a follow-on cleanup to keep the diff here smaller? Reviewers could then easily identify where/if tests have been changed.

File: azure-pipelines.yml
Comment (fangchenli) [inline azure-pipelines.yml:41]: ABI change in Python 3.8, from cp3xm to cp3x.

File: ci/deps/azure-windows-39.yaml
Comment (fangchenli) [inline ci/deps/azure-windows-39.yaml:29]: There are 17 failures in Python 3.9 Windows build relate to `pyarrow`.

Most of them are:
`pyarrow.lib.ArrowNotImplementedError: Support for codec 'snappy' not built`

Stop testing for now. will open a new issue for this.

File: doc/source/whatsnew/v1.4.0.rst
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:111]: if we dont' have an issue to bump this, let's add one.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. prob will wait till after release of 1.3 to merge this one, so if you can periodically rebase would be great.

File: doc/source/whatsnew/v1.4.0.rst
Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.4.0.rst:111]: we have #41329 but is closed. will reopen

Comment (fangchenli) [reply doc/source/whatsnew/v1.4.0.rst:111]: I'll keep 0.17.0 here, bump this later in #41329.

**Comment by simonjayhawkins** [comment]
once we do this, we won't be releasing from master until at least 26 December. Historically, we have had 6 monthly releases (except for 0.22, which was a special case for a api breaking change) so shouldn't be an issue. Are we ready for this?

**Comment by jreback** [comment]
> once we do this, we won't be releasing from master until at least 26 December. Historically, we have had 6 monthly releases (except for 0.22, which was a special case for a api breaking change) so shouldn't be an issue. Are we ready for this?

yep this is fine this early. going to have a look.

File: doc/source/whatsnew/v1.4.0.rst
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.0.rst:83]: my bad. I didn't do this in #42206

```suggestion
| mypy (dev)      | 0.910           |          |         |
```

File: .github/workflows/posix.yml
Comment (simonjayhawkins) [inline .github/workflows/posix.yml:34]: can you retain the "-W error" in numpydev

File: ci/deps/actions-38-db.yaml
Comment (jreback) [inline ci/deps/actions-38-db.yaml:28]: do we need this as 1.18.*?

File: ci/deps/azure-windows-39.yaml
Comment (jreback) [inline ci/deps/azure-windows-39.yaml:19]: not resolving?

File: ci/deps/azure-windows-39.yaml
Comment (jreback) [reply ci/deps/azure-windows-39.yaml:29]: great can you link the issue here

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine to me, a couple of comments, ping on green

File: ci/deps/actions-38-db.yaml
Comment (fangchenli) [reply ci/deps/actions-38-db.yaml:28]: I believe 1.18.* and 1.18 are equivalent. They both resolve to the latest 1.18 version.

**Comment by jreback** [comment]
https://github.com/pandas-dev/pandas/pull/41989/checks?check_run_id=2984645939 random failure?

**Review by jreback — APPROVED** [review]

**Comment by fangchenli** [comment]
> https://github.com/pandas-dev/pandas/pull/41989/checks?check_run_id=2984645939 random failure?

Looks like a flaky test.

**Comment by jreback** [comment]
thanks @fangchenli very nice!

File: requirements-dev.txt
Comment (simonjayhawkins) [inline requirements-dev.txt:70]: @fangchenli not needed with py3.7+?

**Pull request merged**
